File: reference-c/esp32-llm-lab/host_generate.c
// Run the EXACT ESP32-S3 inference (firmware/common/llm.h) on your laptop and
// print a story. No PyTorch, no emulator, no hardware -- just a C compiler.
//   cc -O3 -o gen host_generate.c -lm
//   ./gen model.bin 120
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include "llm.h"
#include "vocab.h"

static uint8_t *read_file(const char *p, size_t *n){
  FILE *f=fopen(p,"rb"); if(!f){perror(p);exit(1);}
  fseek(f,0,SEEK_END); *n=ftell(f); fseek(f,0,SEEK_SET);
  uint8_t *b=malloc(*n); if(fread(b,1,*n,f)!=*n){fprintf(stderr,"short read\n");exit(1);}
  fclose(f); return b;
}
static void emit(int tok){
  if(tok>=VOCAB_N) return;
  const unsigned char *b=VOCAB_BLOB+VOCAB_OFF[tok];
  int len=VOCAB_OFF[tok+1]-VOCAB_OFF[tok];
  fwrite(b,1,len,stdout); fflush(stdout);
}
int main(int argc,char**argv){
  const char *bin = argc>1?argv[1]:"model.bin";
  int N = argc>2?atoi(argv[2]):120;
  size_t n; uint8_t *buf=read_file(bin,&n);
  Model m; if(llm_load(buf,&m)){fprintf(stderr,"bad model magic\n");return 1;}
  int D=m.c.dim,L=m.c.n_layers,P=m.c.ple_dim,F=m.c.ffn,V=m.c.vocab,S=m.c.seq_len;
  fprintf(stderr,"model: V=%d D=%d L=%d H=%d F=%d P=%d  (%.2f MB)\n",
          V,D,L,m.c.n_heads,F,P,n/1e6);
  Scratch s;
  s.x=malloc(D*4); s.h=malloc((F>D?F:D)*4); s.qkv=malloc(3*D*4); s.att=malloc(D*4);
  s.g1=malloc(F*4); s.g2=malloc((P>F?P:F)*4); s.ple=malloc(L*P*4);
  s.tmpP=malloc(L*P*4); s.trow=malloc(L*P*4); s.logits=malloc(V*4); s.scores=malloc(S*4);
  s.kcache=malloc((size_t)L*S*D*4); s.vcache=malloc((size_t)L*S*D*4);
  m.tok_emb.rows = VOCAB_N;                 // skip padded (never-emitted) rows
  int PROMPT[]={336,337,258,338};           // "Once upon a time" (this tokenizer)
  int np=4, pos=0, tok=0;
  printf(">>> ");
  for(int i=0;i<np;i++){ tok=PROMPT[i]; emit(tok); llm_forward(&m,tok,pos++,&s); }
  for(int step=0; step<N && pos<S; step++){
    int best=0; float bv=-1e30f;
    for(int v=0; v<VOCAB_N; v++) if(s.logits[v]>bv){bv=s.logits[v]; best=v;}
    tok=best; emit(tok); llm_forward(&m,tok,pos++,&s);
  }
  printf("\n");
  return 0;
}

File: engine/llm-host/src/bin/host_generate.rs
//! Port of reference-c/esp32-llm-lab/host_generate.c: run the exact chip
//! inference engine on a laptop and print a story. Greedy decode only, no
//! PyTorch, no emulator, no hardware -- just this binary.
//!
//! Usage: host-generate <model.bin> <vocab.h> [N]
//!
//! Deviation from the C original: vocab.h is a runtime argument here
//! instead of a `#include` baked in at compile time (see vocab.rs).
//! Everything else -- the hardcoded prompt, the greedy decode loop, the
//! VOCAB_N-bounded argmax -- matches host_generate.c exactly.

use llm_core::{llm_forward, Model};
use llm_host::{support::ScratchOwned, vocab::Vocab};
use std::fs;
use std::io::Write;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let model_path = args.get(1).map(String::as_str).unwrap_or("model.bin");
    let vocab_path = args.get(2).map(String::as_str).unwrap_or("vocab.h");
    let n: usize = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(120);

    let bytes = fs::read(model_path).unwrap_or_else(|e| {
        eprintln!("{model_path}: {e}");
        std::process::exit(1);
    });
    let model = Model::load(&bytes).unwrap_or_else(|e| {
        eprintln!("bad model magic: {e:?}");
        std::process::exit(1);
    });
    let vocab = Vocab::load(vocab_path).unwrap_or_else(|e| {
        eprintln!("{e}");
        std::process::exit(1);
    });

    let c = &model.cfg;
    eprintln!(
        "model: V={} D={} L={} H={} F={} P={}  ({:.2} MB)",
        c.vocab,
        c.dim,
        c.n_layers,
        c.n_heads,
        c.ffn,
        c.ple_dim,
        bytes.len() as f64 / 1e6
    );

    let mut owned = ScratchOwned::new(&model.cfg);
    let mut s = owned.as_scratch();

    // "Once upon a time" (this tokenizer) -- reference-c/esp32-llm-lab/host_generate.c's
    // own hardcoded prompt array, kept verbatim for parity. (Note: with the
    // vocab.h bundled alongside it in this snapshot, these ids do NOT
    // actually decode to "Once upon a time" -- see MIGRATION_PLAN.md notes.
    // Reproducing the C reference exactly, not "fixing" it, is the point of
    // this binary.)
    let prompt = [336usize, 337, 258, 338];

    let stdout = std::io::stdout();
    let mut out = stdout.lock();
    write!(out, ">>> ").unwrap();

    let mut pos = 0usize;
    // Matches C's `int tok=0;` -- always overwritten before use in both
    // loops below, same as the original. Kept as a plain dead initializer
    // (not restructured into an Option/loop-scoped binding) for structural
    // fidelity to host_generate.c.
    #[allow(unused_assignments)]
    let mut tok = 0usize;
    for &t in &prompt {
        tok = t;
        emit(&mut out, &vocab, tok);
        llm_forward(&model, tok, pos, &mut s);
        pos += 1;
    }
    for _ in 0..n {
        if pos >= model.cfg.seq_len {
            break;
        }
        let mut best = 0usize;
        let mut bv = f32::NEG_INFINITY;
        for v in 0..vocab.n {
            if s.logits[v] > bv {
                bv = s.logits[v];
                best = v;
            }
        }
        tok = best;
        emit(&mut out, &vocab, tok);
        llm_forward(&model, tok, pos, &mut s);
        pos += 1;
    }
    writeln!(out).unwrap();
}

fn emit<W: Write>(out: &mut W, vocab: &Vocab, tok: usize) {
    if let Some(bytes) = vocab.bytes(tok) {
        out.write_all(bytes).ok();
        out.flush().ok();
    }
}
